### ISSUE TITLE
Running CrossEncoder.rank() in batches

### ISSUE THREAD

**Issue by lidija-jovanovska** [UKPLab/sentence-transformers#2812]
Hi! 

I would like to make a feature request for batch inference in the CrossEncoder model to improve performance.

When running `CrossEncoder.rank()`, the user has to specify a query and a list of documents (corpus). With this proposed change the user would specify a list of queries and a list of corresponding corpuses to run the search in.

**Comment by tomaarsen** [comment]
Hello!

I can imagine that there might be some performance improvement possible, but only if you are having a lot of partial batches despite doing all (re)ranking at the same time. I think this is pretty niche - same with running multiple queries on the same, or on different, documents at the same time.
I'm a bit concerned that allowing multiple queries and corpora will make the inputs/outputs of the method a bit more confusing.  So, right now I'm leaning towards not implementing this.

If you are reranking a lot of different queries and corpora simultaneously, then I would recommend creating the query-document pairs yourself and running `model.predict`. Then you can take advantage of the batch size to its full extent.

- Tom Aarsen

**Comment by lidija-jovanovska** [comment]
Hi Tom, thanks for your swift reply :) As you gathered, my use case is to run multiple parallel re-ranking steps on different query-corpus combinations. So, far I was generating the sentence pairs before passing them to `model.predict()`. Recently, I realised that `model.rank()` exists and the endpoint design corresponds much better to my use case, save the lack of support for parallelism. However, I do understand that nesting the request model too much might be more difficult to understand and use. 

Bottom line, I can continue using `model.predict()`. Thanks!
